localstack/services/apigateway/provider.py

```diff
@@ -0,0 +1,701 @@
+    NotFoundException,
+    NullableInteger,
+    RequestValidator,
+    RequestValidators,
+    RestApi,
+    String,
+    Tags,
+    VpcLink,
+    VpcLinks,
+)
+from localstack.aws.proxy import AwsApiListener
+from localstack.constants import HEADER_LOCALSTACK_EDGE_URL
+from localstack.services.apigateway import helpers
+from localstack.services.apigateway.context import ApiInvocationContext
+from localstack.services.apigateway.helpers import (
+    API_REGIONS,
+    PATH_REGEX_TEST_INVOKE_API,
+    PATH_REGEX_USER_REQUEST,
+    APIGatewayRegion,
+    apply_json_patch_safe,
+    find_api_subentity_by_id,
+)
+from localstack.services.apigateway.invocations import invoke_rest_api_from_request
+from localstack.services.moto import call_moto
+from localstack.utils.analytics import event_publisher
+from localstack.utils.aws import aws_stack
+from localstack.utils.aws.aws_responses import requests_response
+from localstack.utils.collections import ensure_list
+from localstack.utils.json import parse_json_or_yaml
+from localstack.utils.strings import short_uid, to_str
+from localstack.utils.time import now_utc
+
+
+class ApigatewayApiListener(AwsApiListener):
+    """Custom API listener that handles both, API Gateway API calls (managing the
+    state/metadata of the service) and invocations (invoking a user-created API)."""
+
+    def forward_request(self, method, path, data, headers):
+        invocation_context = ApiInvocationContext(method, path, data, headers)
+
+        forwarded_for = headers.get(HEADER_LOCALSTACK_EDGE_URL, "")
+        if re.match(PATH_REGEX_USER_REQUEST, path) or "execute-api" in forwarded_for:
+            result = invoke_rest_api_from_request(invocation_context)
+            if result is not None:
+                return result
+
+        if helpers.is_test_invoke_method(method, path):
+            # if call is from test_invoke_api then use http_method to find the integration,
+            #   as test_invoke_api makes a POST call to request the test invocation
+            match = re.match(PATH_REGEX_TEST_INVOKE_API, path)
+            invocation_context.method = match[3]
+            data = parse_json_or_yaml(to_str(data or b""))
+            if data:
+                orig_data = data
+                path_with_query_string = orig_data.get("pathWithQueryString", None)
+                if path_with_query_string:
+                    invocation_context.path_with_query_string = path_with_query_string
+                invocation_context.data = data.get("body")
+                invocation_context.headers = orig_data.get("headers", {})
+            result = invoke_rest_api_from_request(invocation_context)
+            result = {
+                "status": result.status_code,
+                "body": to_str(result.content),
+                "headers": dict(result.headers),
+            }
+            return result
+
+        return super(ApigatewayApiListener, self).forward_request(method, path, data, headers)
```

Comment:
nit; In Python3 you can just `super().forward_request(...)` no?

Reply:
Good catch, updated 👍